core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,451 @@
+use std::collections::BTreeMap;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_primitives::errors::StorageError;
+use near_primitives::receipt::TrieQueueIndices;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::{Gas, ShardId};
+use near_primitives::version::ProtocolFeature;
+use near_schema_checker_lib::ProtocolSchema;
+use near_vm_runner::logic::ProtocolVersion;
+
+use crate::{get, set, TrieUpdate};
+
+use super::receipts_column_helper::TrieQueue;
+use super::TrieAccess;
+
+/// Keeps metadata about receipts stored in the outgoing buffers.
+#[derive(Debug)]
+pub struct OutgoingMetadatas {
+    /// Metadata information for outgoing buffer to each shard.
+    metadatas: BTreeMap<ShardId, ReceiptGroupsQueue>,
+    /// Parameteres which control size of receipt groups.
+    groups_config: ReceiptGroupsConfig,
+}
+
+impl OutgoingMetadatas {
+    /// New empty metadata.
+    pub fn new(groups_config: ReceiptGroupsConfig) -> Self {
+        Self { metadatas: BTreeMap::new(), groups_config }
+    }
+
+    /// Load shard metadata from the trie.
+    /// Returns empty metadata for protocol versions which don't support metadata.
+    /// Make sure to pass shard ids for every shard that has receipts in the outgoing buffer,
+    /// otherwise the metadata for it will be overwritten with empty metadata.
+    pub fn load(
```

Comment:
The code works the way it does because:
* There are already some outgoing buffers, we need to create matching metadata for them
* During resharding a new shard could suddenly appear and we need to initialize the metadata for it
* We can't really scan the trie for existing outgoing buffers because the `TrieAccess` trait doesn't have an iterator

The easiest way to achieve this was to initialize the metadata when a receipt for it appears.

> Can we structure the code in such a way that the load functions are only called with BandwidthScheduler feature enabled?

That's the case right now. `OutgoingMetadatas::load` doesn't do anything when the feature is not enabled. And nothing is pushed/popped from the metadata before protocol change.

> In which case, can we get rid of all logic that assumes optional loading? Examples are update_* functions where we do .entry().or_insert_with() and ReceiptGroupsQueue::load() returning Option<Self>?

During resharding there could be a new shard, the `or_insert_with` is meant to handle this case.

> Initialization on enabling BandwidthScheduler feature could be handled by checking if trie key exists and returning default.

That is kind of what's happening, `OutgoingMetadatas::load` tries to load the metadata, and if the trie doesn't contain the key it creates a new one.


But I'm open to other designs as well, it's a bit convoluted :/

Reply:
> That's the case right now. OutgoingMetadatas::load doesn't do anything when the feature is not enabled. And nothing is pushed/popped from the metadata before protocol change.

Oh I see, the calls to `update_on_receipt_pushed` and `update_on_receipt_popped` are guarded by `receipt.should_update_outgoing_metadatas()`, hmm.... That's not at all intuitive

I was about to ask how does the initialization of OutgoingMetadatas on protocol change, but I think I see what's going on. So, I guess we can continue to process V0 stored receipts (that don't update metadata) while creating and processing V1 stored receipts after the protocol upgrade?